# SQL Part 3 Exercises Joins

This notebook works with the BooksDB dataset.  

**Questions 1-4** focus on how joins work in a more general sense.  If you are not sure how to best answer the questions, try manipulating the queries by changing the order of joins, the length of the TOP, or how they are ordered.  

> <span style="background-color: rgba(127, 127, 127, 0.1);"><i>Note</i>: If some of the queries are not returning quickly, you can try closing and restarting azure data studio after a few minutes.</span>

In **questions 5-8**, you will be asked to join tables to find results about titles, tags, and the books users want to read.

As you go through the questions, you will see 'Minimum Desired Output'.  This will provide you with the bare minimum columns needed to answer the questions.  When you write the queries, add as many additional columns as you need to best find your answers.

## **QUESTION 1:**  Exploring a LEFT JOIN using the **books** and **to\_read** tables

The code block below has been commented out. To answer this question, you will need to uncomment it by removing the **/\*** (line 1) and **\*/** (line 16).

**Part A:** Look at Query 1, which table is the 'Left' table and which is the 'Right'?

> Your Answer: Left table is to\_read ; Right table is Books

**Part B:** Using table 2, answer the following questions: 

- When we added a  '**WHERE tr.user\_id IS NOT NULL'** clause (line 14), how did the table change?  Why?

> Your Answer: The query got rid of user\_ids that presented NULL values

In [37]:
select top 10 *
FROM BooksDB.dbo.books 

(10 rows affected)

Total execution time: 00:00:00.338

id book_id best_book_id work_id books_count isbn isbn13 authors original_publication_year original_title title language_code average_rating ratings_count work_ratings_count work_text_reviews_count ratings_1 ratings_2 ratings_3 ratings_4 ratings_5 image_url small_image_url 1 2767052 2767052 2792775 272 439023483 9780439023480 Suzanne Collins 2008 The Hunger Games The Hunger Games (The Hunger Games, #1) eng 4.34 4780653 4942365 155254 66715 127936 560092 1481305 2706317 https://images.gr-assets.com/books/1447303603m/2767052.jpg https://images.gr-assets.com/books/1447303603s/2767052.jpg 2 3 3 4640799 491 439554934 9780439554930 J.K. Rowling, Mary GrandPré 1997 Harry Potter and the Philosopher's Stone Harry Potter and the Sorcerer's Stone (Harry Potter, #1) eng 4.44 4602479 4800065 75867 75504 101676 455024 1156318 3011543 https://images.gr-assets.com/books/1474154022m/3.jpg https://images.gr-assets.com/books/1474154022s/3.jpg 3 41865 41865 3212258 226 316015849 9780316015840 Stephenie Meyer 2005 Twilight Twilight (Twilight, #1) en-US 3.57 3866839 3916824 95009 456191 436802 793319 875073 1355439 https://images.gr-assets.com/books/1361039443m/41865.jpg https://images.gr-assets.com/books/1361039443s/41865.jpg 4 2657 2657 3275794 487 61120081 9780061120080 Harper Lee 1960 To Kill a Mockingbird To Kill a Mockingbird eng 4.25 3198671 3340896 72586 60427 117415 446835 1001952 1714267 https://images.gr-assets.com/books/1361975680m/2657.jpg https://images.gr-assets.com/books/1361975680s/2657.jpg 5 4671 4671 245494 1356 743273567 9780743273560 F. Scott Fitzgerald 1925 The Great Gatsby The Great Gatsby eng 3.89 2683664 2773745 51992 86236 197621 606158 936012 947718 https://images.gr-assets.com/books/1490528560m/4671.jpg https://images.gr-assets.com/books/1490528560s/4671.jpg 6 11870085 11870085 16827462 226 525478817 9780525478810 John Green 2012 The Fault in Our Stars The Fault in Our Stars eng 4.26 2346404 2478609 140739 47994 92723 327550 698471 1311871 https://images.gr-assets.com/books/1360206420m/11870085.jpg https://images.gr-assets.com/books/1360206420s/11870085.jpg 7 5907 5907 1540236 969 618260307 9780618260300 J.R.R. Tolkien 1937 The Hobbit or There and Back Again The Hobbit en-US 4.25 2071616 2196809 37653 46023 76784 288649 665635 1119718 https://images.gr-assets.com/books/1372847500m/5907.jpg https://images.gr-assets.com/books/1372847500s/5907.jpg 8 5107 5107 3036731 360 316769177 9780316769170 J.D. Salinger 1951 The Catcher in the Rye The Catcher in the Rye eng 3.79 2044241 2120637 44920 109383 185520 455042 661516 709176 https://images.gr-assets.com/books/1398034300m/5107.jpg https://images.gr-assets.com/books/1398034300s/5107.jpg 9 960 960 3338963 311 1416524797 9781416524790 Dan Brown 2000 Angels & Demons Angels & Demons (Robert Langdon, #1) en-CA 3.85 2001311 2078754 25112 77841 145740 458429 716569 680175 https://images.gr-assets.com/books/1303390735m/960.jpg https://images.gr-assets.com/books/1303390735s/960.jpg 10 1885 1885 3060926 3455 679783261 9780679783270 Jane Austen 1813 Pride and Prejudice Pride and Prejudice eng 4.24 2035490 2191465 49152 54700 86485 284852 609755 1155673 https://images.gr-assets.com/books/1320399351m/1885.jpg https://images.gr-assets.com/books/1320399351s/1885.jpg

In [38]:

-- Query 1: LEFT JOIN 
SELECT TOP 50 b.title, b.average_rating, b.books_count,tr.user_id, tr.book_id
FROM BooksDB.dbo.books AS b
LEFT JOIN BooksDB.dbo.to_read AS tr
ON b.best_book_id = tr.book_id
ORDER BY b.average_rating;

-- Query 2: LEFT JOIN USING 'IS NOT NULL'
SELECT TOP 50 b.title, b.average_rating, b.books_count,tr.user_id, tr.book_id
FROM BooksDB.dbo.books AS b
LEFT JOIN BooksDB.dbo.to_read AS tr
ON b.best_book_id = tr.book_id
WHERE tr.user_id IS NOT NULL
ORDER BY b.average_rating;


(50 rows affected)

(50 rows affected)

Total execution time: 00:00:07.367

title average_rating books_count user_id book_id One Night at the Call Center 2.47 25 NULL NULL The Almost Moon 2.67 63 NULL NULL The Finkler Question 2.76 45 NULL NULL Four Blondes 2.8 57 449 6613 Four Blondes 2.8 57 1378 6613 Four Blondes 2.8 57 6465 6613 Four Blondes 2.8 57 6700 6613 Four Blondes 2.8 57 7830 6613 Four Blondes 2.8 57 10085 6613 Four Blondes 2.8 57 16298 6613 Four Blondes 2.8 57 16845 6613 Four Blondes 2.8 57 17470 6613 Four Blondes 2.8 57 19763 6613 Four Blondes 2.8 57 21938 6613 Four Blondes 2.8 57 22289 6613 Four Blondes 2.8 57 23758 6613 Four Blondes 2.8 57 27874 6613 Four Blondes 2.8 57 28690 6613 Four Blondes 2.8 57 29023 6613 Four Blondes 2.8 57 29315 6613 Four Blondes 2.8 57 30245 6613 Four Blondes 2.8 57 36128 6613 Four Blondes 2.8 57 36554 6613 Four Blondes 2.8 57 38159 6613 Four Blondes 2.8 57 38800 6613 Four Blondes 2.8 57 40361 6613 Four Blondes 2.8 57 40529 6613 Four Blondes 2.8 57 41485 6613 Four Blondes 2.8 57 42706 6613 Four Blondes 2.8 57 45206 6613 Four Blondes 2.8 57 45689 6613 Four Blondes 2.8 57 49499 6613 Four Blondes 2.8 57 51105 6613 Four Blondes 2.8 57 51671 6613 Four Blondes 2.8 57 52109 6613 Lost 2.8 18 NULL NULL Tell-All 2.84 37 NULL NULL Revenge Wears Prada: The Devil Returns (The Devil Wears Prada, #2) 2.84 59 NULL NULL Among the Ten Thousand Things 2.84 19 NULL NULL The Emperor's Children 2.93 45 7860 3648 The Emperor's Children 2.93 45 11320 3648 The Emperor's Children 2.93 45 12686 3648 The Emperor's Children 2.93 45 12729 3648 The Emperor's Children 2.93 45 12891 3648 The Emperor's Children 2.93 45 16933 3648 The Emperor's Children 2.93 45 17093 3648 The Emperor's Children 2.93 45 24036 3648 The Emperor's Children 2.93 45 35122 3648 The Emperor's Children 2.93 45 36022 3648 The Emperor's Children 2.93 45 36952 3648

title average_rating books_count user_id book_id Four Blondes 2.8 57 449 6613 Four Blondes 2.8 57 1378 6613 Four Blondes 2.8 57 6465 6613 Four Blondes 2.8 57 6700 6613 Four Blondes 2.8 57 7830 6613 Four Blondes 2.8 57 10085 6613 Four Blondes 2.8 57 16298 6613 Four Blondes 2.8 57 16845 6613 Four Blondes 2.8 57 17470 6613 Four Blondes 2.8 57 19763 6613 Four Blondes 2.8 57 21938 6613 Four Blondes 2.8 57 22289 6613 Four Blondes 2.8 57 23758 6613 Four Blondes 2.8 57 27874 6613 Four Blondes 2.8 57 28690 6613 Four Blondes 2.8 57 29023 6613 Four Blondes 2.8 57 29315 6613 Four Blondes 2.8 57 30245 6613 Four Blondes 2.8 57 36128 6613 Four Blondes 2.8 57 36554 6613 Four Blondes 2.8 57 38159 6613 Four Blondes 2.8 57 38800 6613 Four Blondes 2.8 57 40361 6613 Four Blondes 2.8 57 40529 6613 Four Blondes 2.8 57 41485 6613 Four Blondes 2.8 57 42706 6613 Four Blondes 2.8 57 45206 6613 Four Blondes 2.8 57 45689 6613 Four Blondes 2.8 57 49499 6613 Four Blondes 2.8 57 51105 6613 Four Blondes 2.8 57 51671 6613 Four Blondes 2.8 57 52109 6613 The Emperor's Children 2.93 45 7860 3648 The Emperor's Children 2.93 45 11320 3648 The Emperor's Children 2.93 45 12686 3648 The Emperor's Children 2.93 45 12729 3648 The Emperor's Children 2.93 45 12891 3648 The Emperor's Children 2.93 45 16933 3648 The Emperor's Children 2.93 45 17093 3648 The Emperor's Children 2.93 45 24036 3648 The Emperor's Children 2.93 45 35122 3648 The Emperor's Children 2.93 45 36022 3648 The Emperor's Children 2.93 45 36952 3648 The Emperor's Children 2.93 45 38430 3648 The Emperor's Children 2.93 45 39414 3648 The Emperor's Children 2.93 45 45261 3648 The Emperor's Children 2.93 45 45497 3648 The Emperor's Children 2.93 45 45747 3648 The Emperor's Children 2.93 45 48502 3648 The Emperor's Children 2.93 45 49233 3648

## **QUESTION 2:** Exploring **RIGHT** joins with the **books** and **to\_read** tables

You will need to uncomment the Query to answer the question.

**Part A:** Looking at the query, which table is the 'Left' and which is the 'Right'?

> Your Answer: Right table is to\_read ; Left table is Books

**Part B:** How do the RIGHT-joined tables differ from the LEFT-joined tables in Question 1?  

> Your Answer:

In [39]:

-- Query 1: RIGHT JOIN
SELECT TOP 300 b.title, b.average_rating, b.books_count, tr.user_id, tr.book_id
FROM BooksDB.dbo.books AS b
RIGHT JOIN BooksDB.dbo.to_read AS tr
ON b.best_book_id = tr.book_id;


(300 rows affected)

Total execution time: 00:00:00.076

title average_rating books_count user_id book_id Children of Dune (Dune Chronicles #3) 3.9 116 1 112 NULL NULL NULL 1 235 NULL NULL NULL 1 533 NULL NULL NULL 1 1198 NULL NULL NULL 1 1874 NULL NULL NULL 1 2058 NULL NULL NULL 1 3334 NULL NULL NULL 2 4 NULL NULL NULL 2 11 The Ultimate Hitchhiker's Guide to the Galaxy 4.37 32 2 13 NULL NULL NULL 2 16 NULL NULL NULL 2 92 Dune Messiah (Dune Chronicles #2) 3.86 140 2 106 NULL NULL NULL 2 113 NULL NULL NULL 2 233 NULL NULL NULL 2 247 NULL NULL NULL 2 380 NULL NULL NULL 2 408 NULL NULL NULL 2 439 NULL NULL NULL 2 479 NULL NULL NULL 2 516 NULL NULL NULL 2 619 NULL NULL NULL 2 1069 The Iliad 3.83 1726 2 1371 NULL NULL NULL 2 1384 NULL NULL NULL 2 1597 NULL NULL NULL 2 1926 NULL NULL NULL 2 1967 NULL NULL NULL 2 1969 NULL NULL NULL 2 2220 NULL NULL NULL 2 2388 NULL NULL NULL 2 3222 NULL NULL NULL 2 3227 NULL NULL NULL 2 3733 NULL NULL NULL 2 5418 NULL NULL NULL 2 6937 NULL NULL NULL 2 7146 NULL NULL NULL 2 7873 NULL NULL NULL 2 8417 Harry Potter Collection (Harry Potter, #1-6) 4.73 6 3 10 NULL NULL NULL 3 66 NULL NULL NULL 3 233 NULL NULL NULL 3 726 NULL NULL NULL 3 932 NULL NULL NULL 3 1146 NULL NULL NULL 3 2439 NULL NULL NULL 3 3566 NULL NULL NULL 3 4582 NULL NULL NULL 3 5448 NULL NULL NULL 3 5765 NULL NULL NULL 3 6274 NULL NULL NULL 3 8403 NULL NULL NULL 5 31 The Virtue of Selfishness: A New Concept of Egoism 3.47 53 5 665 NULL NULL NULL 5 3074 Sherlock Holmes: The Complete Novels and Stories, Volume I 4.45 13 5 3581 NULL NULL NULL 6 64 NULL NULL NULL 6 459 NULL NULL NULL 6 619 NULL NULL NULL 6 649 NULL NULL NULL 6 687 NULL NULL NULL 6 874 The War of Art: Break Through the Blocks & Win Your Inner Creative Battles 4.07 29 6 1319 NULL NULL NULL 6 1912 NULL NULL NULL 6 5269 NULL NULL NULL 6 5537 NULL NULL NULL 6 5820 NULL NULL NULL 6 6220 NULL NULL NULL 6 7480 Harry Potter Collection (Harry Potter, #1-6) 4.73 6 7 10 NULL NULL NULL 7 215 NULL NULL NULL 7 225 NULL NULL NULL 7 233 NULL NULL NULL 7 1322 NULL NULL NULL 7 3552 NULL NULL NULL 7 3720 NULL NULL NULL 7 4468 NULL NULL NULL 8 19 NULL NULL NULL 8 498 NULL NULL NULL 8 1357 NULL NULL NULL 8 2177 NULL NULL NULL 8 3930 The Pilgrimage 3.64 171 8 4004 NULL NULL NULL 8 8792 NULL NULL NULL 8 9304 Harry Potter Boxed Set, Books 1-5 (Harry Potter, #1-5) 4.77 11 9 8 Children of Dune (Dune Chronicles #3) 3.9 116 9 112 NULL NULL NULL 9 139 NULL NULL NULL 9 230 NULL NULL NULL 9 233 NULL NULL NULL 9 473 NULL NULL NULL 9 638 NULL NULL NULL 9 653 NULL NULL NULL 9 672 NULL NULL NULL 9 701 NULL NULL NULL 9 736 NULL NULL NULL 9 883 NULL NULL NULL 9 895 NULL NULL NULL 9 1467 NULL NULL NULL 9 1582 NULL NULL NULL 9 1873 NULL NULL NULL 9 2516 NULL NULL NULL 9 2630 NULL NULL NULL 9 3210 Icy Sparks 3.68 27 9 3476 NULL NULL NULL 9 3560 NULL NULL NULL 9 3650 NULL NULL NULL 9 5286 NULL NULL NULL 9 6636 Harry Potter and the Goblet of Fire (Harry Potter, #4) 4.53 332 10 6 NULL NULL NULL 10 385 Atlas Shrugged 3.68 147 10 662 NULL NULL NULL 10 896 NULL NULL NULL 10 1076 NULL NULL NULL 10 1082 NULL NULL NULL 10 1220 NULL NULL NULL 10 1281 NULL NULL NULL 10 1439 NULL NULL NULL 10 1851 NULL NULL NULL 10 1920 NULL NULL NULL 10 2441 NULL NULL NULL 10 2570 NULL NULL NULL 10 2616 NULL NULL NULL 10 3059 NULL NULL NULL 10 3545 NULL NULL NULL 10 3961 NULL NULL NULL 10 4369 NULL NULL NULL 10 4510 NULL NULL NULL 10 5104 NULL NULL NULL 10 5254 NULL NULL NULL 10 5556 NULL NULL NULL 10 6114 NULL NULL NULL 10 6917 NULL NULL NULL 10 7545 NULL NULL NULL 10 7842 NULL NULL NULL 11 3643 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 12 1 NULL NULL NULL 12 124 NULL NULL NULL 12 1471 NULL NULL NULL 12 2893 NULL NULL NULL 13 81 NULL NULL NULL 13 277 NULL NULL NULL 13 318 NULL NULL NULL 13 430 NULL NULL NULL 13 437 NULL NULL NULL 13 551 NULL NULL NULL 13 560 NULL NULL NULL 13 610 NULL NULL NULL 13 678 NULL NULL NULL 13 979 NULL NULL NULL 13 1082 NULL NULL NULL 13 1464 NULL NULL NULL 13 1577 NULL NULL NULL 13 2018 NULL NULL NULL 13 2418 NULL NULL NULL 13 3257 NULL NULL NULL

## **QUESTION 3:** Exploring FULL JOINS with the **books** and **to\_read** tables

You will need to uncomment the Query to answer the question.

**Part A:**  Look at the table and explore how it was populated.  Try manipulating the query to better understand how this table works.  Try adding a WHERE clause, or ORDER BY a column on the **books** table.  Based on your exploration, can you think of when you might want a FULL join rather than a LEFT, RIGHT, or INNER?

> Your Answer: You would use a full join whenever you want to seee all the data laid out, including all null values, all data from all tables.

In [40]:

-- FULL JOIN
SELECT TOP 30 b.title, b.average_rating, b.books_count,tr.user_id, tr.book_id
FROM BooksDB.dbo.books AS b
FULL JOIN BooksDB.dbo.to_read AS tr
ON b.book_id = tr.book_id;


(30 rows affected)

Total execution time: 00:00:00.088

title average_rating books_count user_id book_id Children of Dune (Dune Chronicles #3) 3.9 116 1 112 NULL NULL NULL 1 235 NULL NULL NULL 1 533 NULL NULL NULL 1 1198 NULL NULL NULL 1 1874 NULL NULL NULL 1 2058 NULL NULL NULL 1 3334 NULL NULL NULL 2 4 The Hitchhiker's Guide to the Galaxy (Hitchhiker's Guide to the Galaxy, #1) 4.2 257 2 11 The Ultimate Hitchhiker's Guide to the Galaxy 4.37 32 2 13 NULL NULL NULL 2 16 NULL NULL NULL 2 92 Dune Messiah (Dune Chronicles #2) 3.86 140 2 106 NULL NULL NULL 2 113 NULL NULL NULL 2 233 NULL NULL NULL 2 247 NULL NULL NULL 2 380 NULL NULL NULL 2 408 NULL NULL NULL 2 439 NULL NULL NULL 2 479 NULL NULL NULL 2 516 NULL NULL NULL 2 619 NULL NULL NULL 2 1069 The Iliad 3.83 1726 2 1371 NULL NULL NULL 2 1384 NULL NULL NULL 2 1597 NULL NULL NULL 2 1926 NULL NULL NULL 2 1967 NULL NULL NULL 2 1969 NULL NULL NULL 2 2220

In [41]:
 -- Right Join
SELECT TOP 30 b.title, b.average_rating, b.books_count,tr.user_id, tr.book_id
FROM BooksDB.dbo.books AS b
FULL JOIN BooksDB.dbo.to_read AS tr
ON b.book_id = tr.book_id
WHERE b.title is not NULL AND b.average_rating is not NULL AND b.books_count is NOT NULL AND tr.user_id is NOT NULL AND tr.book_id IS NOT NULL
ORDER BY b.title;




(30 rows affected)

Total execution time: 00:00:08.044

title average_rating books_count user_id book_id 1421: The Year China Discovered America 3.58 40 3884 4813 1421: The Year China Discovered America 3.58 40 4452 4813 1421: The Year China Discovered America 3.58 40 6421 4813 1421: The Year China Discovered America 3.58 40 8273 4813 1421: The Year China Discovered America 3.58 40 9127 4813 1421: The Year China Discovered America 3.58 40 20535 4813 1421: The Year China Discovered America 3.58 40 24039 4813 1421: The Year China Discovered America 3.58 40 27571 4813 1421: The Year China Discovered America 3.58 40 29254 4813 1421: The Year China Discovered America 3.58 40 30890 4813 1421: The Year China Discovered America 3.58 40 31048 4813 1421: The Year China Discovered America 3.58 40 34073 4813 1421: The Year China Discovered America 3.58 40 34131 4813 1421: The Year China Discovered America 3.58 40 38973 4813 1421: The Year China Discovered America 3.58 40 42026 4813 1421: The Year China Discovered America 3.58 40 43050 4813 1776 4.06 38 673 1067 1776 4.06 38 3427 1067 1776 4.06 38 4461 1067 1776 4.06 38 5703 1067 1776 4.06 38 5786 1067 1776 4.06 38 6537 1067 1776 4.06 38 7911 1067 1776 4.06 38 8183 1067 1776 4.06 38 8353 1067 1776 4.06 38 8554 1067 1776 4.06 38 10206 1067 1776 4.06 38 10544 1067 1776 4.06 38 11068 1067 1776 4.06 38 11275 1067

## **QUESTION 4:** Exploring INNER Joins with the **books** and **to\_read** tables

You will need to uncomment the Query to answer the question.

**Part A:**  What does an inner join do?

> <span style="background-color: rgba(127, 127, 127, 0.1); font-family: -apple-system, BlinkMacSystemFont, sans-serif;">Your Answer:&nbsp;</span> 
> 
> <span style="background-color: rgba(127, 127, 127, 0.1); font-family: -apple-system, BlinkMacSystemFont, sans-serif;">Inner join selects all rows from both the tables as long as the condition is satisfied. This keyword will create the result by combining all rows from both the tables where the condition satisdies. The value of the common field is the same.&nbsp;</span> 

<span style="background-color: rgba(127, 127, 127, 0.1); font-family: -apple-system, BlinkMacSystemFont, sans-serif;"><b>Part B:</b>&nbsp;Currently this table has an <b>'IS NOT NULL'</b> in the <b>WHERE</b> clause.&nbsp; Is this necessary for this table?&nbsp; Why or Why not?&nbsp;</span> 

> Your Answer: If I am understanding correctly, inner join will only give a result where there are values , so it would not give a result where there are vulll values in one field and not null values in the field next to it. 

**Part C:**  When using SQL, you are usually returning much larger tables so every line of code will add to the compilation and return time.  Rewrite this inner join query so that it contains no redundant code.   

> Your Answer:  Update the query in the code box below.

In [42]:

-- INNER JOIN
SELECT TOP 30 b.title, b.average_rating, b.books_count,tr.user_id, tr.book_id
FROM BooksDB.dbo.books AS b
INNER JOIN BooksDB.dbo.to_read AS tr
ON b.book_id = tr.book_id
ORDER BY tr.book_id;


(30 rows affected)

Total execution time: 00:00:00.134

title average_rating books_count user_id book_id Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 12 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 192 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 202 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 244 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 253 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 350 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 478 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 511 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 525 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 527 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 552 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 644 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 754 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 826 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 1043 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 1205 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 1442 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 1470 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 1712 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 1814 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 1919 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 2051 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 2052 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 2158 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 2160 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 2345 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 2476 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 2523 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 2623 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 2673 1

## **QUESTION 5:**  Using joins to find the most tagged book and the most popular tag

We will be writing 3 queries to find both the most tagged book and that book's most popular tag.

**Part A:**  What is the most tagged book?

> <span style="background-color: rgba(127, 127, 127, 0.1); font-family: -apple-system, BlinkMacSystemFont, sans-serif;">Start by joining the </span> **books** <span style="background-color: rgba(127, 127, 127, 0.1); font-family: -apple-system, BlinkMacSystemFont, sans-serif;">and </span> **book\_tags**        <span style="background-color: rgba(127, 127, 127, 0.1); font-family: -apple-system, BlinkMacSystemFont, sans-serif;">&nbsp;tables ON the books.best_book_id and book_tags.goodreads_book_id.&nbsp;&nbsp;<br></span>              <span style="background-color: rgba(127, 127, 127, 0.1); font-family: -apple-system, BlinkMacSystemFont, sans-serif;">We want the <i>most popular</i> book, so think about ordering the table in a way that will display both the book title and the number of times a book tag has been used.</span>
> 
> _MInimum Desired output:_  The title of the most tagged book, and the number of times the book has been tagged.
> 
> **Title:**  Harry Potter

**Part B:** How many different tags have been used for the most tagged book? **100**

> <span style="font-family: -apple-system, BlinkMacSystemFont, sans-serif; background-color: rgba(127, 127, 127, 0.1);">Start by joining the&nbsp;</span>        **books** <span style="font-family: -apple-system, BlinkMacSystemFont, sans-serif; background-color: rgba(127, 127, 127, 0.1);">and&nbsp;</span>        **book\_tags**        <span style="font-family: -apple-system, BlinkMacSystemFont, sans-serif; background-color: rgba(127, 127, 127, 0.1);">&nbsp;tables ON the books.best_book_id and book_tags.goodreads_book_id.&nbsp;&nbsp;<br></span>              We want to find the _most popular_ tag id, so think about ordering in a way that will display the most popular tag id.
> 
> <span style="font-family: -apple-system, BlinkMacSystemFont, sans-serif; background-color: rgba(127, 127, 127, 0.1);"><i>Minimum Desired Output:</i> The tag id for the most popular tag used for the most tagged book.&nbsp;</span>       
> 
> <span style="font-family: -apple-system, BlinkMacSystemFont, sans-serif; background-color: rgba(127, 127, 127, 0.1);"><b>Tag Id:&nbsp;</b></span>       30574

<span style="font-family: -apple-system, BlinkMacSystemFont, sans-serif; background-color: rgba(127, 127, 127, 0.1);"><b>Part C: </b>What is this particular tag?</span>

> You may return this answer with a simple query.

> <span style="font-family: -apple-system, BlinkMacSystemFont, sans-serif; background-color: rgba(127, 127, 127, 0.1);"><i>Minimum Desired Output:</i>&nbsp;The tag name for the most popular tag used for the most tagged book.&nbsp;</span>       
> 
> <span style="font-family: -apple-system, BlinkMacSystemFont, sans-serif; background-color: rgba(127, 127, 127, 0.1);"><b>Tag Name: to_read&nbsp;</b></span>

In [57]:
-- Code Here: MOST TAGGED BOOK
SELECT TOP 5 
SUM (count) 'tag_count', b.best_book_id, title
FROM BooksDB.dbo.books AS b
LEFT JOIN BooksDB.dbo.book_tags AS bt
ON b.best_book_id = bt.goodreads_book_id
GROUP BY b.best_book_id, title
ORDER BY tag_count DESC;


Warning: Null value is eliminated by an aggregate or other SET operation.

(5 rows affected)

Total execution time: 00:00:02.432

tag_count best_book_id title 786374 3 Harry Potter and the Sorcerer's Stone (Harry Potter, #1) 679219 2429135 The Girl with the Dragon Tattoo (Millennium, #1) 675971 865 The Alchemist 600025 18143977 All the Light We Cannot See 597465 24280 Les Misérables

In [58]:
-- NUMBER OF TAGS THE MOST TAGGED BOOK HAS
SELECT 
COUNT (tag_id) 'Number of tags'
FROM book_tags
WHERE goodreads_book_id = 3;

(1 row affected)

Total execution time: 00:00:00.158

Number of tags 100

In [59]:
-- NUMBER OF TIMES MOST POPULAR TAG WAS USED FOR MOST POPULAR BOOK
SELECT TOP 5 
bt.count, tag_id, b.title, bt.goodreads_book_id
FROM books AS b
LEFT JOIN book_tags AS bt
ON b.best_book_id = bt.goodreads_book_id
WHERE b.best_book_id = 3
Order BY bt.count DESC;

(5 rows affected)

Total execution time: 00:00:00.120

count tag_id title goodreads_book_id 496107 30574 Harry Potter and the Sorcerer's Stone (Harry Potter, #1) 3 48220 11557 Harry Potter and the Sorcerer's Stone (Harry Potter, #1) 3 47478 11305 Harry Potter and the Sorcerer's Stone (Harry Potter, #1) 3 44640 8717 Harry Potter and the Sorcerer's Stone (Harry Potter, #1) 3 14984 33114 Harry Potter and the Sorcerer's Stone (Harry Potter, #1) 3

In [60]:
-- MOST POPULAR TAG NAME USED ON MOST POPULAR BOOK
SELECT TOP 5 *
FROM tags
WHERE tag_id = 30574;

(1 row affected)

Total execution time: 00:00:00.054

tag_id tag_name 30574 to-read

## **QUESTION 6:** Find a book based on a tag of your choice using joins.

Some suggestions: any hobbies you like? baking, drawing, etc. how about favorite holidays, subjects, animals, etc?

**Part A:** Explore the tags table for a tag of your choice.  Suggested ideas, try searching for one of your hobbies, a favorite time of year, favorite animal, etc.  Start by querying the tag table for your chosen tag.  (_Hint:_ in order to make sure you find all combinations of your tags, use the wildcard).  If your query returns more than one result, pick your favorite.

> > _Here is an example.  Alyce wanted to search for halloween books.  She used the wildcard, '%halloween%' and 8 different combinations were found.  Alyce selected '**halloween-storytime**', with a tag id of 13877._

> _Minimum Desired Output:_ <span style="font-family: -apple-system, BlinkMacSystemFont, sans-serif; background-color: rgba(127, 127, 127, 0.1);">tag name and tag id</span>

> **Your Tag and Tag Id: tag ID= 22098 ; Tag name = nutrition**

**Part B:** Now that you have the **tag id**, find the title that was tagged the most with this tag by joining the books and books\_tags tables.

> > _Alyce joined the book table with the tags table and discovered that <u>The Little Old Lady Who Was Not Afraid of Anything</u> was tagged as '**halloween-storytime**' the most._

> _Minimum Desired Output:_ <span style="background-color: rgba(127, 127, 127, 0.1); font-family: -apple-system, BlinkMacSystemFont, sans-serif;">title and tag count</span>

> **Your Title and Number of times the tag was used:** 
> 
> **The Omnivore's Dilemma: A Natural History of Four Meals - tagged 275 times**

In [61]:
-- Code Here: NUTRITION BOOKS
Select top 5 bt.goodreads_book_id, bt.tag_id, t.tag_name
From book_tags AS bt
LEFT JOIN tags AS t
ON bt.tag_id = t.tag_id
WHERE tag_name = 'nutrition'
ORDER BY count DESC;

(5 rows affected)

Total execution time: 00:00:00.133

goodreads_book_id tag_id tag_name 3109 22098 nutrition 315425 22098 nutrition 11505008 22098 nutrition 7015635 22098 nutrition 8727466 22098 nutrition

In [62]:
-- TITLE TAGGED THE MOST AS 'NUTRITION, 22098'
SELECT TOP 5
SUM (bt.count) AS tag_count, title
FROM books AS b
LEFT JOIN book_tags AS bt
ON b.best_book_id = bt.goodreads_book_id
WHERE tag_id = 22098
GROUP BY title
ORDER BY tag_count DESC;

(5 rows affected)

Total execution time: 00:00:00.146

tag_count title 275 The Omnivore's Dilemma: A Natural History of Four Meals 261 In Defense of Food: An Eater's Manifesto 149 Wheat Belly: Lose the Wheat, Lose the Weight, and Find Your Path Back to Health 142 Food Rules: An Eater's Manual 139 Why We Get Fat: And What to Do About It

## **QUESTION 7:** What are the top 10 most popular book titles on the 'to read' list?

**Part A:**  Explore how the **to\_read** and **books** tables are organized, and then join them.  What will you use as a key?

> Your Answer:

**Part B:** Organize your newly joined table in a way that groups popularity based on the title based on users.

> _Hint:_ Suggest using an alias for the aggregation you will need to run on the **user\_id**

> _Minimum Desired Output:_ title and aliased column

In [47]:
-- Code Here:

Commands completed successfully.

Total execution time: 00:00:00.043

In [64]:
SELECT top 10
COUNT (tr.book_id) 'most popular book count', tr.book_id, title
FROM tp_read AS tr
LEFT JOIN books AS b
ON tr.book_id = b.book_id
WHERE b.title IS NOT NULL
GROUP BY tr.book_id, b.title
ORDER BY (most popular book count) DESC;

: Msg 102, Level 15, State 1, Line 8
Incorrect syntax near 'popular'.

Total execution time: 00:00:00.041

## **QUESTION 8:** Largest 'To Read' Lists

**Part A:**  Create a query that returns a table that contains the top 10 users with the most number of titles on their 'to read' list.    Group all the entries by user\_id.  

> _Hint:_ You might want an alias for the titles.
> 
> _Minimum Desired Output:_  user\_id and your aliased column of titles.

**Part B:**  The longest list length is 15 titles total, and is shared by 4 different users.  Select one of the uses and print out their entire '**to read**' list.

In [48]:
-- Code Here:

Commands completed successfully.

Total execution time: 00:00:00.042